### ISSUE TITLE
LinkedIn integration issues

### ISSUE THREAD

**Issue by ahaider48** [FormidableLabs/react-native-app-auth#200]
Hello,

I am trying to integrate **Linkedin** to work with this library and I am having a few issues.

The project I am working on, requires me to get the `accessToken` from LinkedIn and send it to my backend to be exchanged for another token.

Linkedin has their own native [Android](https://developer.linkedin.com/docs/android-sdk-auth) and [iOS](https://developer.linkedin.com/docs/ios-sdk-auth) sdk which they recommend that mobile users integrate, but the issue with these are that:
1) The access token can only be used via the sdk and can't be passed to the backend to make requests 
2) It requires the mobile user to download the LinkedIn app locally first (i.e. no login via a webpage!)

As such, I am forced to use their normal [Web Oauth 2](https://developer.linkedin.com/docs/oauth2) flow, which uses the web form to login and the access token retrieved can be used anywhere.

The only issue is that LinkedIn doesn't allow url schemes (e.g. `com.site.app:/oauth2redirect`) to be used as call back urls. 

To get around this, on my node/express backend I created a special call back route (e.g. `https://site.com/app/oauth2redirect`) that just redirects any requests it receives along with all the query params (`token`, `state`, etc) to my app scheme URL (e.g. `com.site.app:/oauth2redirect`). I provided this url to Linkedin as my allowed redirect url and used it to configure the auth request on the app.

This approach worked in that on my iOS/Android, the redirect was happening properly (I tried launching the maps app to test this and it worked!). However, the promise to `authorize()` was not being resolved after the redirect.

I am not sure why the call just hangs. I can successfully retrieve the "authorization code" but for whatever reason, ~~I feel that the library doesn't see the app loading via the url scheme after the in-app browser closes and whatever function that is supposed to be called then is not triggered!~~ I think that maybe my redirect doesn't behave the way a direct redirect from auth provider does. In android, I get the error message `Failed exchange token` (iOS no error) which probably means that my URL scheme/intent in the format of `com.site.app:/oauth2redirect?code&state` isn't picked up the same way as others!

I tried also using Universal Links on iOS instead and I observed a similar issue, albeit, in this approach, I feel that the way universal links are handled are different than url scheme and hence that's why it wasn't working!

I was wondering if anyone has tried any of these two approaches with LinkedIn and knows why the promise doesn't resolve after the redirect. Maybe I am doing something wrong or I need to modify my `appDelegate.m` for this use case? (especially for universal link approach!)

I am trying to make it work with url schemes and the redirect. I followed all the instructions in the setup. I can successfully authenticate with Google on iOS/Android. So my setup looks fine. Only LinkedIn authentication fails.

This is my callback function/config for the button:

```js
onLinkedinLogin = async () => {
    const config = {
      serviceConfiguration: {
        authorizationEndpoint: 'https://www.linkedin.com/oauth/v2/authorization',
        tokenEndpoint: 'https://www.linkedin.com/oauth/v2/accessToken',
      },
      clientId: CLIENT_ID,
      clientSecret: CLIENT_SECRET,
      redirectUrl: 'http://localhost:8080/app/oauth2redirect', // on the server this redirects to `com.site.app:/oauth2redirect`
      scopes: ['r_basicprofile', 'r_emailaddress'],
    };

    try {
      const { accessToken } = await authorize(config);
      console.log('accessToken', accessToken);
    } catch (err) {
      console.error(err);
    }
  }
```

A few things to note:

- For now, I am okay with using the secret here (like I said the sdk generated tokens can't be used in the backend)
- I already have Google authentication set up and its working on both iOS and Android
- The `authorize()` promise never resloves after the in-app browser closes after a successful login in iOS and in Android I get the error `Failed exchange token`.

Can this be done with this library? Any alternatives/issues with my approach?

Any help/comment with issue issue would be appriciated. 

P.S. I wish we could just have a way to get the authorization code only with this library and let the backend exchange it for the token as suggested [here](https://github.com/FormidableLabs/react-native-app-auth/issues/96).

**Comment by ahaider48** [comment]
As an update, I think the cause of my issue is related to https://github.com/openid/AppAuth-iOS/issues/232

**Comment by Petemir** [comment]
@ahaider48 I'm sorry to interfere but I have one question. You mention that you have Google authentication set up and its working, does it follow the same flow? Redirecting the user to your backend, and from there using deeplinks/universal links to your app? Because I am trying that and I cannot do it because I don't have the PKCE code_verifier in the backend to make the exchange with the authorization code for the access token...

Thanks!

**Comment by ahaider48** [comment]
@Petemir Sorry, no. I am using this approach *only* for LinkedIn (with deep links/url schemes). I couldn't use this library for LinkedIn.

For Google, I am just using whatever was suggested by this library to get the `accessToken` and I send it to the backend for any other auth work!

**Comment by Rajesh7890** [comment]
Can anyone have the idea of how to resolve authorize() when redirect to a server url

**Comment by Rajesh7890** [comment]
I can but I am not getting the auth data from the server. It stays there @andresubri

**Comment by sabaakmal023** [comment]
> Hello,
> 
> I am trying to integrate **Linkedin** to work with this library and I am having a few issues.
> 
> The project I am working on, requires me to get the `accessToken` from LinkedIn and send it to my backend to be exchanged for another token.
> 
> Linkedin has their own native [Android](https://developer.linkedin.com/docs/android-sdk-auth) and [iOS](https://developer.linkedin.com/docs/ios-sdk-auth) sdk which they recommend that mobile users integrate, but the issue with these are that:
> 
> 1. The access token can only be used via the sdk and can't be passed to the backend to make requests
> 2. It requires the mobile user to download the LinkedIn app locally first (i.e. no login via a webpage!)
> 
> As such, I am forced to use their normal [Web Oauth 2](https://developer.linkedin.com/docs/oauth2) flow, which uses the web form to login and the access token retrieved can be used anywhere.
> 
> The only issue is that LinkedIn doesn't allow url schemes (e.g. `com.site.app:/oauth2redirect`) to be used as call back urls.
> 
> To get around this, on my node/express backend I created a special call back route (e.g. `https://site.com/app/oauth2redirect`) that just redirects any requests it receives along with all the query params (`token`, `state`, etc) to my app scheme URL (e.g. `com.site.app:/oauth2redirect`). I provided this url to Linkedin as my allowed redirect url and used it to configure the auth request on the app.
> 
> This approach worked in that on my iOS/Android, the redirect was happening properly (I tried launching the maps app to test this and it worked!). However, the promise to `authorize()` was not being resolved after the redirect.
> 
> I am not sure why the call just hangs. I can successfully retrieve the "authorization code" but for whatever reason, ~I feel that the library doesn't see the app loading via the url scheme after the in-app browser closes and whatever function that is supposed to be called then is not triggered!~ I think that maybe my redirect doesn't behave the way a direct redirect from auth provider does. In android, I get the error message `Failed exchange token` (iOS no error) which probably means that my URL scheme/intent in the format of `com.site.app:/oauth2redirect?code&state` isn't picked up the same way as others!
> 
> I tried also using Universal Links on iOS instead and I observed a similar issue, albeit, in this approach, I feel that the way universal links are handled are different than url scheme and hence that's why it wasn't working!
> 
> I was wondering if anyone has tried any of these two approaches with LinkedIn and knows why the promise doesn't resolve after the redirect. Maybe I am doing something wrong or I need to modify my `appDelegate.m` for this use case? (especially for universal link approach!)
> 
> I am trying to make it work with url schemes and the redirect. I followed all the instructions in the setup. I can successfully authenticate with Google on iOS/Android. So my setup looks fine. Only LinkedIn authentication fails.
> 
> This is my callback function/config for the button:
> 
> ```js
> onLinkedinLogin = async () => {
>     const config = {
>       serviceConfiguration: {
>         authorizationEndpoint: 'https://www.linkedin.com/oauth/v2/authorization',
>         tokenEndpoint: 'https://www.linkedin.com/oauth/v2/accessToken',
>       },
>       clientId: CLIENT_ID,
>       clientSecret: CLIENT_SECRET,
>       redirectUrl: 'http://localhost:8080/app/oauth2redirect', // on the server this redirects to `com.site.app:/oauth2redirect`
>       scopes: ['r_basicprofile', 'r_emailaddress'],
>     };
> 
>     try {
>       const { accessToken } = await authorize(config);
>       console.log('accessToken', accessToken);
>     } catch (err) {
>       console.error(err);
>     }
>   }
> ```
> 
> A few things to note:
> 
> * For now, I am okay with using the secret here (like I said the sdk generated tokens can't be used in the backend)
> * I already have Google authentication set up and its working on both iOS and Android
> * The `authorize()` promise never resloves after the in-app browser closes after a successful login in iOS and in Android I get the error `Failed exchange token`.
> 
> Can this be done with this library? Any alternatives/issues with my approach?
> 
> Any help/comment with issue issue would be appriciated.
> 
> P.S. I wish we could just have a way to get the authorization code only with this library and let the backend exchange it for the token as suggested [here](https://github.com/FormidableLabs/react-native-app-auth/issues/96).

any suggestion to resolve this issues?

**Comment by Jay-A-McBee** [comment]
I was able to get this working using a universal link on iOS, but this library is not well-suited to using LinkedIn as an OAuth provider. You must include the `client_secret` with the request to `https://www.linkedin.com/oauth/v2/accessToken` meaning you would have to include this value in your provisioned binary. This opens the app up to an "Authorization Code Interception Attack" where a malicious app could intercept the redirect, gain access to the `client_secret` and successfully exchange the authorization code for an access token.

As the OP points out attempting to redirect to a backend endpoint that will then redirect to your app's custom url scheme won't work either. The value passed to the the native AppAuth libraries as the redirectUrl (ex. 'http://localhost:8080/app/oauth2redirect') is used to determine whether or not those libs actually handle the incoming response. See [AppAuth-ios shouldHandleUrl fn](https://github.com/openid/AppAuth-iOS/blob/e191b1beadf3041259652e6e464eaceff16e5fbb/Source/AppAuthCore/OIDAuthorizationService.m#L119). So attempting to redirect to a custom url scheme that wasn't passed as the `redirectUrl` simply won't work on iOS.

Closing this out as we are unable to support this provider.
